source/library/Argo/Codec.hs

```diff
@@ -0,0 +1,284 @@
+module Argo.Codec where
+
+import Control.Applicative ((<|>))
+
+import qualified Argo.Json.Array as Array
+import qualified Argo.Json.Boolean as Boolean
+import qualified Argo.Json.Member as Member
+import qualified Argo.Json.Name as Name
+import qualified Argo.Json.Null as Null
+import qualified Argo.Json.Number as Number
+import qualified Argo.Json.Object as Object
+import qualified Argo.Json.String as String
+import qualified Argo.Json.Value as Value
+import qualified Argo.Result as Result
+import qualified Control.Applicative as Applicative
+import qualified Control.Monad as Monad
+import qualified Control.Monad.Trans.Class as Trans
+import qualified Control.Monad.Trans.Except as Except
+import qualified Control.Monad.Trans.Reader as Reader
+import qualified Control.Monad.Trans.State as State
+import qualified Control.Monad.Trans.Writer as Writer
+import qualified Data.Text as Text
+
+decodeWith :: ValueCodec a -> Value.Value -> Result.Result a
+decodeWith c = either fail pure . Except.runExcept . Reader.runReaderT (decode c)
+
+encodeWith :: ValueCodec a -> a -> Value.Value
+encodeWith c x = State.execState (encode c x) . Value.Null $ Null.fromUnit ()
+
+project :: (i -> f) -> CodecOf r w f o -> CodecOf r w i o
+project f c = c { encode = encode c . f }
+
+data CodecOf r w i o = Codec
+    { decode :: r o
+    , encode :: i -> w o
+    }
+
+instance (Functor r, Functor w) => Functor (CodecOf r w i) where
+    fmap f c = Codec
+        { decode = fmap f $ decode c
+        , encode = fmap f . encode c
+        }
+
+instance (Applicative r, Applicative w) => Applicative (CodecOf r w i) where
+    pure x = Codec
+        { decode = pure x
+        , encode = const $ pure x
+        }
+    cf <*> cx = Codec
+        { decode = decode cf <*> decode cx
+        , encode = \ i -> encode cf i <*> encode cx i
+        }
+
+instance (Applicative.Alternative r, Applicative.Alternative w) => Applicative.Alternative (CodecOf r w i) where
+    empty = Codec
+        { decode = Applicative.empty
+        , encode = const Applicative.empty
+        }
+    cx <|> cy = Codec
+        { decode = decode cx <|> decode cy
+        , encode = \ i -> encode cx i <|> encode cy i
+        }
+
+type Codec r w a = CodecOf r w a a
+
+dimap :: (Functor r, Functor w) => (a -> b) -> (b -> a) -> Codec r w a -> Codec r w b
+dimap f g c = Codec
+    { decode = fmap f $ decode c
+    , encode = fmap f . encode c . g
+    }
+
+type ValueCodec a = Codec
+    (Reader.ReaderT Value.Value (Except.Except String))
```

Comment:
Perhaps this (and others) should use [`Errors`](https://hackage.haskell.org/package/transformers-0.6.0.2/docs/Control-Applicative-Lift.html#t:Errors) instead of `Except` to allow for multiple simultaneous errors?

Reply:
Oh, `Errors` does not have a `Monad` instance. I'd have to use something like [`Validation`](https://hackage.haskell.org/package/validation-1.1.1/docs/Data-Validation.html#t:Validation) instead.